redis/src/cluster_topology.rs

```diff
@@ -61,24 +62,59 @@ impl TopologyView {
 struct SlotMapValue {
     start: u16,
     addrs: SlotAddrs,
+    latest_used_replica: AtomicUsize,
 }
 
 impl SlotMapValue {
     fn from_slot(slot: Slot) -> Self {
         Self {
             start: slot.start(),
             addrs: SlotAddrs::from_slot(slot),
+            latest_used_replica: AtomicUsize::new(0),
         }
     }
 }
 
-#[derive(Debug, Default)]
-pub(crate) struct SlotMap(BTreeMap<u16, SlotMapValue>);
+#[derive(Debug, Default, Clone, PartialEq, Copy)]
+pub(crate) enum ReadFromReplicaStrategy {
+    #[default]
+    AlwaysFromPrimary,
+    RoundRobin,
+}
+
+#[derive(Debug)]
+pub(crate) struct SlotMap {
+    slots: BTreeMap<u16, SlotMapValue>,
+    read_from_replica: ReadFromReplicaStrategy,
+}
+
+fn get_address_from_slot(
+    slot: &SlotMapValue,
+    read_from_replica: ReadFromReplicaStrategy,
+    slot_addr: SlotAddr,
+) -> &str {
+    if slot_addr == SlotAddr::Master || slot.addrs.replicas.is_empty() {
+        return slot.addrs.primary.as_str();
+    }
+    match read_from_replica {
+        ReadFromReplicaStrategy::AlwaysFromPrimary => slot.addrs.primary.as_str(),
+        ReadFromReplicaStrategy::RoundRobin => {
+            let index = slot
```

Comment:
if all replicas failed, but still in the replicas addresses, do we have a fallback to the primary?

Reply:
Good question. 

```rs
if slot_addr == SlotAddr::Master || slot.addrs.replicas.is_empty() {
        return slot.addrs.primary.as_str();
    }
```

ensures that if the `SlotMap` doesn't know about the replica, then the command will go to a primary. But if the `ConnectionMap` has removed a replica, the slot map isn't updated about this. This will cause the SlotMap to return a connection that the connection map doesn't know about, which will cause the cluster connection to send the request to a random node. problem.

Sounds like we can't keep the separation between connection and slot map.

Reply:
Another scenario is if the replicas are present in the CLUSTER SLOTS output, but they aren't responsive (e.g. due to load or still wasn't recognized as failed in the cluster bus). Or, if the replicas have been filtered out from the cluster bus, but still in our connection map/slot map since we haven't done slots refreshment yet. So in both options we will face connection errors / timeouts.

So, basically-  what i'm asking is how the retries mechanisms of a read from replica command is implemented?
How do we handle a timeout/connection error when querying a replica? do we need to wait and retry to connect to this replica? Do we route it to the next replica? When do we decide to send it to the primary - only if no replicas are present?

I think we should try querying the next node immediately. We should try to reconnect on the background, and/or start a topology refresh after several of connection errors, but as we have more nodes options to send this command to, I don't think we should delay the command execution.

Reply:
I think that those decisions are about removing a connection from the connections manager. we can do that, but its beyond the scope of this PR.

Reply:
> connection error redis-rs blocks this command on trying to reconnect to this node and then retrying to send the command to the same node

AFAIS on every retry the cluster connection requests a new connection from the container, so if there are multiple replicas, the request will be rotated. The only case where the connection is specified is when using `CommandRouting::Connection`, which is only created on multi-node requests. When using `CommandRouting::Route` the route is resolved on each retry.